### PR TITLE
Master mdl 71378 90

### PR DESCRIPTION
**Pull request by marcghaly**
PR for Issue #90 : Helper class with methods : get_populated and course_exists + unit tests

### CODE REVIEW

File: mod/qbank/classes/helper.php [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+     * @param int $categoryid Category id parent to the course
+     * @return bool Returns a boolean.
+     * @throws \moodle_exception
+     */
+    public static function create_category_course(string $coursename, int $categoryid): stdClass {
+        $newcourse = new stdClass();
+
+        $newcourse->fullname = $coursename;
+        $newcourse->category = $categoryid;
+        create_course($newcourse);
```
Comment (mattporritt) [inline mod/qbank/classes/helper.php:83]: I'm not sure how this doesn't explode?
The create_course function in Moodle doesn't check if the course shortname variable exists, but the DB table has it as a requirement: https://github.com/moodle/moodle/blob/master/lib/db/install.xml#L76
And in the Moodle UI shortname is required.
We should set a short name, or at least have it passed in as a variable.

File: mod/qbank/classes/helper.php [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+     * Creates course.
+     *
+     * @param string $coursename Course name to create
+     * @param int $categoryid Category id parent to the course
+     * @return bool Returns a boolean.
+     * @throws \moodle_exception
+     */
+    public static function create_category_course(string $coursename, int $categoryid): stdClass {
+        $newcourse = new stdClass();
+
```
Comment (mattporritt) [inline mod/qbank/classes/helper.php:80]: Lets add a course description here from a lang string that will help users understand what and why this copurse has been created.
Some like (but better): 
"""
This course has been created as part of the upgrade to Moodle 4.0.
The question bank in this category has been migrated to a question bank activity in this course.
All the questions that were previously located in this categories question bank have been migrated to the question bank activity in this course.
"""

File: mod/qbank/db/install.php [hunk 3]
```diff
@@ -23,12 +23,26 @@
...
 function xmldb_qbank_install() {
 
-    return true;
+    // Get data from populated questions.
+    $questioncategories = helper::get_categories_populated();
+    foreach ($questioncategories as $qcategory) {
+        $contextlevel = (int)$qcategory->contextlevel;
+
+        if ($contextlevel === CONTEXT_COURSECAT) {
+            $coursename = substr("qbank_" . CONTEXT_COURSECAT . $qcategory->name, 0, 254);
```
Comment (mattporritt) [inline mod/qbank/db/install.php:41]: I know I gave zero direction here, But we should come up with a more human friendly course name.
Maybe "Question bank: . $qcategory->name" still with the trimming

File: mod/qbank/classes/helper.php (hunk 1 above)
Comment (marcghaly) [reply mod/qbank/classes/helper.php:83]: Added a shortname passed as variable `$newcourse->shortname = $shortname;`

File: mod/qbank/classes/helper.php (hunk 2 above)
Comment (marcghaly) [reply mod/qbank/classes/helper.php:80]: Added description in lang file `$string['qbank_migratedcourse_description']` assigned in `create_category_course method`

File: mod/qbank/db/install.php (hunk 3 above)
Comment (marcghaly) [reply mod/qbank/db/install.php:41]: Modified `$coursename` as asked

**Comment by guillogo** [comment]
Thank you @mattporritt for reviewing this PR. the changes were done. Also, I've added some changes to prepare code for issue 91 and clean everything in 1 commit.

**Review by mattporritt — APPROVED** [review]

**Pull request merged**